File: src/python/challenges/problems/sort_matrix_by_diagonals_test.py

# https://leetcode.com/problems/sort-matrix-by-diagonals

from typing import List
from core.problem_base import *

class SortMatrixByDiagonals(ProblemBase):
    def Solution(self, grid: List[List[int]]) -> List[List[int]]:
        n = len(grid)

        diagonals = [[] for _ in range(2 * n - 1)]

        for row in range(n):
            for col in range(n):
                diagonal = n + col - row - 1
                diagonals[diagonal].append(grid[row][col])

        for i in range(len(diagonals)):
            if i < n:
                diagonals[i].sort(reverse=True)
            else:
                diagonals[i].sort()

            for j in range(len(diagonals[i])):
                if i < n:
                    row = n - len(diagonals[i]) + j
                    col = j
                else:
                    row = j
                    col = n - len(diagonals[i]) + j
                grid[row][col] = diagonals[i][j]

        return grid

if __name__ == '__main__':
    TestGen(SortMatrixByDiagonals) \
        .Add(lambda tc: tc.Param([[1,7,3,5],[9,8,2,2],[4,5,6,8],[4,2,8,7]]).Result([[8,2,2,5],[9,7,7,3],[4,8,6,8],[4,2,5,1]]) ) \
        .Add(lambda tc: tc.Param([[1,7,3],[9,8,2],[4,5,6]]).Result([[8,2,3],[9,6,7],[4,5,1]]) ) \
        .Add(lambda tc: tc.Param([[0,1],[1,2]]).Result([[2,1],[1,0]]) ) \
        .Run()

File: src/go/challenges/problems/sort_matrix_by_diagonals_test.go
// https://leetcode.com/problems/sort-matrix-by-diagonals

package problems

import (
	"sort"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type SortMatrixByDiagonals struct{}

func TestSortMatrixByDiagonals(t *testing.T) {
	gen := core.TestSuite[SortMatrixByDiagonals]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([][]int{{1, 7, 3, 5}, {9, 8, 2, 2}, {4, 5, 6, 8}, {4, 2, 8, 7}}).Result([][]int{{8, 2, 2, 5}, {9, 7, 7, 3}, {4, 8, 6, 8}, {4, 2, 5, 1}})
	}).Run(t)
}

func (SortMatrixByDiagonals) Solution(grid [][]int) [][]int {
	n := len(grid)

	diagonals := [][]int{}

	for i := 0; i < 2*n-1; i++ {
		diagonals = append(diagonals, []int{})
	}

	for row := 0; row < n; row++ {
		for col := 0; col < n; col++ {
			diagonal := n + col - row - 1
			diagonals[diagonal] = append(diagonals[diagonal], grid[row][col])
		}
	}

	for i := 0; i < len(diagonals); i++ {
		if i < n {
			sort.Sort(sort.Reverse(sort.IntSlice(diagonals[i])))
		} else {
			sort.Ints(diagonals[i])
		}

		length := len(diagonals[i])
		for j := 0; j < length; j++ {
			row := n - length + j
			col := j
			if i >= n {
				row = j
				col = n - length + j
			}

			grid[row][col] = diagonals[i][j]
		}
	}

	return grid
}
